device/src/ota.rs

```diff
@@ -0,0 +1,521 @@
+            .unwrap_or(0)
+    }
+
+    pub fn active_partition(&self, flash: &mut FlashStorage) -> Partition {
+        match self.current_slot(flash) {
+            Some((slot, _, _)) => self.ota_partitions[slot as usize],
+            None => self.factory_partition,
+        }
+    }
+
+    pub fn current_slot(&self, flash: &mut FlashStorage) -> Option<(u8, u32, Option<OtaMetadata>)> {
+        let mut best_partition = None;
+        let mut best_seq = 0;
+        let mut best_metadata = None;
+        for slot in 0u8..=1 {
+            let mut bytes = [0u8; (ESP32_OTADATA_SIZE + FS_PARTITION_METADATA_SIZE) as usize];
+            flash
+                .read(self.otadata_offset + slot as u32 * SECTOR_SIZE, &mut bytes)
+                .unwrap();
+            let seq = u32::from_le_bytes(bytes[0..4].try_into().unwrap());
+            let implied_cs = CRC.checksum(&seq.to_le_bytes());
+            let actual_cs = u32::from_le_bytes(bytes[28..28 + 4].try_into().unwrap());
+            let metadata = bincode::decode_from_slice::<OtaMetadata, _>(
+                &bytes[ESP32_OTADATA_SIZE as usize..],
+                bincode::config::standard(),
+            )
+            .ok()
+            .map(|(metadata, _)| metadata);
+            if implied_cs != actual_cs {
+                continue;
+            }
+
+            if best_partition.is_none() || seq >= best_seq {
+                best_partition = Some(slot);
+                best_seq = seq;
+                best_metadata = metadata;
+            }
+        }
+
+        let current_partition = best_partition?;
+        Some((current_partition, best_seq, best_metadata))
+    }
+
+    pub fn read_metadata_in_slot(&self, flash: &mut FlashStorage, slot: u8) -> Option<OtaMetadata> {
+        let mut bytes = [0u8; FS_PARTITION_METADATA_SIZE as usize];
+        flash
+            .read(
+                self.otadata_offset + (slot as u32) * SECTOR_SIZE + ESP32_OTADATA_SIZE,
+                &mut bytes[..],
+            )
+            .unwrap();
+        if bytes != [0xff; FS_PARTITION_METADATA_SIZE as usize] {
+            let (metadata, _) =
+                bincode::decode_from_slice(&bytes[..], bincode::config::standard()).ok()?;
+            Some(metadata)
+        } else {
+            None
+        }
+    }
+
+    fn switch_partition(&self, flash: &mut FlashStorage, slot: u8, metadata: Option<OtaMetadata>) {
+        let next_seq = match self.current_slot(flash) {
+            Some((current_slot, seq, _)) => {
+                if slot == current_slot {
+                    /* do nothing */
+                    return;
+                } else {
+                    seq + 1
+                }
+            }
+            None => 1,
+        };
+
+        let cs = CRC.checksum(&next_seq.to_le_bytes());
+        let mut bytes = [0xffu8; ESP32_OTADATA_SIZE as usize + FS_PARTITION_METADATA_SIZE as usize];
+        bytes[0..4].copy_from_slice(&next_seq.to_le_bytes());
```

Comment:
The purpose of this checksum is to guarantee that we never load into an old partition if these 'sequence' bytes were corrupted?

Reply:
Yes. This is just what the default bootloader wants so that's what it gets.